model/flow/protocol_state.go

```diff
@@ -0,0 +1,203 @@
+		}
+
+		nextEpochProtocolState := protocolState.NextEpochProtocolState
+		nextEpochIdentityTable, err := buildIdentityTable(
+			nextEpochProtocolState.Identities,
+			nextEpochSetup.Participants,
+			currentEpochSetup.Participants,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("could not build next epoch identity table: %w", err)
+		}
+
+		// fill identities for next epoch
+		result.NextEpochProtocolState = &RichProtocolStateEntry{
+			ProtocolStateEntry:     nextEpochProtocolState,
+			CurrentEpochSetup:      nextEpochSetup,
+			CurrentEpochCommit:     nextEpochCommit,
+			PreviousEpochSetup:     result.CurrentEpochSetup,  // previous epoch setup is current epoch setup
+			PreviousEpochCommit:    result.CurrentEpochCommit, // previous epoch setup is current epoch setup
+			Identities:             nextEpochIdentityTable,
+			NextEpochProtocolState: nil, // always nil
+		}
+	} else {
+		// if next epoch is not yet created, it means that we are in staking phase,
+		// so we need to build the identity table using previous and current epoch setup events.
+		var otherIdentities IdentityList
+		if previousEpochSetup != nil {
+			otherIdentities = previousEpochSetup.Participants
+		}
+		result.Identities, err = buildIdentityTable(
+			protocolState.Identities,
+			currentEpochSetup.Participants,
+			otherIdentities,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("could not build identity table for staking phase: %w", err)
+		}
+	}
+
+	return result, nil
+}
+
+// ID returns hash of entry by hashing all fields.
+func (e *ProtocolStateEntry) ID() Identifier {
+	if e == nil {
+		return ZeroID
+	}
+	body := struct {
+		CurrentEpochEventIDs            Identifier
+		PreviousEpochEventIDs           Identifier
+		Identities                      DynamicIdentityEntryList
+		InvalidStateTransitionAttempted bool
+		NextEpochProtocolStateID        Identifier
+	}{
+		CurrentEpochEventIDs:            e.CurrentEpochEventIDs.ID(),
+		PreviousEpochEventIDs:           e.PreviousEpochEventIDs.ID(),
+		Identities:                      e.Identities,
+		InvalidStateTransitionAttempted: e.InvalidStateTransitionAttempted,
+		NextEpochProtocolStateID:        e.NextEpochProtocolState.ID(),
+	}
+	return MakeID(body)
+}
+
+// Sorted returns whether the list is sorted by the input ordering.
+func (ll DynamicIdentityEntryList) Sorted(less IdentifierOrder) bool {
+	for i := 0; i < len(ll)-1; i++ {
+		a := ll[i]
+		b := ll[i+1]
+		if !less(a.NodeID, b.NodeID) {
+			return false
+		}
+	}
+	return true
+}
+
+// buildIdentityTable builds identity table for current epoch combining data from previous, current epoch setups and dynamic identities
+// that are stored in protocol state. It also performs sanity checks to make sure that data is consistent.
+// No errors are expected during normal operation.
+func buildIdentityTable(
+	dynamicIdentities DynamicIdentityEntryList,
+	coreIdentities, otherIdentities IdentityList,
```

Comment:
I'd like to propose some more precise terminology to use here. Existing code working with these concepts has had issues with a lack of precise terminology for the different states -- I think there's a good opportunity to improve that here.

I wrote up an extension to one of the epoch docs here: https://www.notion.so/dapperlabs/Epoch-Participant-Lifecycle-0c7b7a6ecd904a9fb8c023069488fe51?pvs=4, which enumerates some names for all the possible states.

Reply:
```suggestion
	authParticipantIdentities, authNonParticipantIdentities IdentityList,
```

Reply:
It's an improvement over what was there, IMO. But I still think there is space for explicitly naming each of the possible states. It's already covered in https://github.com/onflow/flow-go/issues/4649